lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,219 @@
+ */
+
+#include "glow/Optimizer/GraphOptimizer/GraphOptimizer.h"
+
+#include "glow/Backend/Backend.h"
+#include "glow/Graph/Graph.h"
+#include "glow/Graph/Log.h"
+#include "glow/Graph/Node.h"
+#include "glow/Graph/Nodes.h"
+#include "glow/Graph/PlaceholderBindings.h"
+#include "glow/Optimizer/GraphOptimizer/FunctionPasses.h"
+
+#include "llvm/Support/Casting.h"
+
+using namespace glow;
+using llvm::cast;
+using llvm::dyn_cast;
+using llvm::isa;
+
+namespace {
+/// The name of the temporary function to be used to perform constant folding.
+constexpr const char *constEvaluationFunctionName =
+    "__constEvaluationFunction__";
+
+/// \returns true if a node \p N is a constant operation, i.e. it is a constant
+/// or all of its inputs are constants, and it has no side-effects and supported
+/// by the \p backend.
+bool isConstantOperation(const Node *N, const Backend &backend) {
+  // An operation with side-effects cannot be computed at compile-time.
+  if (N->hasSideEffects()) {
+    return false;
+  }
+  // Constant and splat nodes are trivially constant operations.
+  if (llvm::isa<Constant>(N) || llvm::isa<SplatNode>(N)) {
+    return true;
+  }
+  // If the operation is not supported by the backend, it cannot be computed at
+  // compile-time.
+  if (!backend.isOpSupported(NodeInfo(*N))) {
+    return false;
+  }
+  if (llvm::isa<Placeholder>(N)) {
+    return false;
+  }
+  for (unsigned idx = 0, e = N->getNumInputs(); idx < e; ++idx) {
+    auto input = N->getNthInput(idx);
+    if (!isConstantOperation(input.getNode(), backend)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+/// Compile the function \p F for the provided \p backend using the compilation
+/// context \p cctx. \returns compiled function
+std::unique_ptr<CompiledFunction> compile(Backend &backend, Function &F,
+                                          CompilationContext &cctx) {
+  EXIT_ON_ERR(::glow::optimizeFunction(&F, backend, cctx));
+  for (const Node &N : F.getNodes()) {
+    CHECK(backend.isOpSupported(N))
+        << "Backend must support all nodes after high-level optimizations but "
+           "encountered unsupported operator: "
+        << N.getDebugDesc();
+  }
+  auto funcOrErr = backend.compile(&F, cctx.backendOpts);
+  EXIT_ON_ERR(funcOrErr.takeError());
+  return std::move(*funcOrErr);
+}
+
+/// Runs the compiled function \p compiledF from the module \p M on the \p
+/// backend using provided \p bindings.
+void run(Backend &backend, CompiledFunction &compiledF, Function &F,
+         PlaceholderBindings &bindings) {
+  std::unique_ptr<PlaceholderBindings> bindingsPtr(&bindings);
+  ExecutionContext context(std::move(bindingsPtr));
+  // Allocate only placeholders used by F.
+  for (auto ph : F.findPlaceholders()) {
+    // Don't allocate tensors for placeholders that are already allocated.
+    if (context.getPlaceholderBindings()->count(ph)) {
+      continue;
+    }
+    context.getPlaceholderBindings()->allocate(ph);
```

Comment:
Why is this necessary? Didn't we already allocate all of the save nodes' PHs? And so there shouldn't be any more PHs otherwise right?

Reply:
I guess I'm a bit overprotective here. IIRC, I was thinking that eventually this "simplified" way to run functions on a backend may be used elsewhere, where it is not guaranteed that all placeholders are allocated already.

But I can remove this logic for now if you think it overcomplicates things.